src/util/animator.js

```diff
@@ -14,33 +14,60 @@ var Element = function(id) {
   this.y = 0;
   this.id = id;
   this.angle = 0;
-  this.render = function(canvas) {
+};
+
+Element.prototype = {
+  render: function(canvas) {
     throw "Rendering abstract element not implemented.";
-  };
+  },
+};
+
+// An animation is a change to an element's property during some time period.
+// fn = function(t, e, initial) that gives the property's value at time t,
+// given the current state of the element (e) and its initial state (initial).
+// The element's initial state will only contain primitive properties
+// (more specifically, everything that can be serialized in JSON).
+var Animation = function(start_time, end_time, element_id, property, fn) {
+  this.start_time = start_time;
+  this.end_time = end_time;
+  this.element_id = element_id;
+  this.property = property;
+  this.fn = fn;
+  this.initial_state = null;
+};
+
+Animation.prototype = {
+  // Returns whether the animation is active at time `t`.
+  isPlayingAt: function(t) {
+    return this.start_time <= t && this.end_time >= t;
+  },
 };
 
 // An element that is a colored rectangle.
 var RectangleElement = function(id, width, height, color, stroke_color,
                                 line_width) {
   var rectangle = this;
   Element.apply(rectangle, [id]);
+
   this.width = width || 0;
   this.height = height || 0;
   this.color = color || 'black';
   this.stroke_color = stroke_color || 'black';
   this.line_width = line_width || 0;
+}
 
-  this.render = function(canvas) {
+RectangleElement.prototype = Object.assign(Element.prototype, {
```

Comment:
This will replace the implementation of render in Element's prototype with the one defined here, which is not wanted. Should be replaced with:
```
RectangleElement.prototype = Object.create(Element.prototype);
Object.assign(RectangleElement.prototype, {
```